presto-common/src/main/java/com/facebook/presto/common/Page.java

```diff
@@ -0,0 +1,450 @@
+            this.blocks = EMPTY_BLOCKS;
+            // Empty blocks are not considered "retained" by any particular page
+            this.retainedSizeInBytes = INSTANCE_SIZE;
+        }
+        else {
+            this.blocks = blocksCopyRequired ? blocks.clone() : blocks;
+        }
+    }
+
+    public int getChannelCount()
+    {
+        return blocks.length;
+    }
+
+    public int getPositionCount()
+    {
+        return positionCount;
+    }
+
+    public long getSizeInBytes()
+    {
+        long sizeInBytes = this.sizeInBytes;
+        if (sizeInBytes < 0) {
+            sizeInBytes = 0;
+            for (Block block : blocks) {
+                sizeInBytes += block.getSizeInBytes();
+            }
+            this.sizeInBytes = sizeInBytes;
+        }
+        return sizeInBytes;
+    }
+
+    public long getLogicalSizeInBytes()
+    {
+        long logicalSizeInBytes = this.logicalSizeInBytes;
+        if (logicalSizeInBytes < 0) {
+            logicalSizeInBytes = 0;
+            for (Block block : blocks) {
+                logicalSizeInBytes += block.getLogicalSizeInBytes();
+            }
+            this.logicalSizeInBytes = logicalSizeInBytes;
+        }
+        return logicalSizeInBytes;
+    }
+
+    /**
+     * Returns the approximate logical size of the page if logicalSizeInBytes was not calculated before.
+     */
+    public long getApproximateLogicalSizeInBytes()
+    {
+        if (logicalSizeInBytes < 0) {
+            long approximateLogicalSizeInBytes = 0;
+            for (Block block : blocks) {
+                approximateLogicalSizeInBytes += block.getApproximateRegionLogicalSizeInBytes(0, block.getPositionCount());
+            }
+            return approximateLogicalSizeInBytes;
+        }
+
+        return logicalSizeInBytes;
+    }
+
+    public long getRetainedSizeInBytes()
+    {
+        long retainedSizeInBytes = this.retainedSizeInBytes;
+        if (retainedSizeInBytes < 0) {
+            return updateRetainedSize();
+        }
+        return retainedSizeInBytes;
+    }
+
+    public Block getBlock(int channel)
+    {
+        return blocks[channel];
+    }
+
+    /**
+     * Gets the values at the specified position as a single element page.  The method creates independent
+     * copy of the data.
+     */
+    public Page getSingleValuePage(int position)
+    {
+        Block[] singleValueBlocks = new Block[this.blocks.length];
+        for (int i = 0; i < this.blocks.length; i++) {
+            singleValueBlocks[i] = this.blocks[i].getSingleValueBlock(position);
+        }
+        return wrapBlocksWithoutCopy(1, singleValueBlocks);
+    }
+
+    public Page getRegion(int positionOffset, int length)
+    {
+        if (positionOffset < 0 || length < 0 || positionOffset + length > positionCount) {
+            throw new IndexOutOfBoundsException(format("Invalid position %s and length %s in page with %s positions", positionOffset, length, positionCount));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: positionOffset (declared at line 163), positionCount, length (declared at line 163). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 166. Consider formatting this data with &#39;%d&#39; instead.